Added unique return codes

fra87 opened the pull request:
Added return codes as anticipated in #884

The return codes are now an Enum inheriting from str, so an explicative message can be automatically printed without helper functions.

The return codes are in file [returncodes.py](https://github.com/jmcnamara/XlsxWriter/compare/main...fra87:returnCodesToEnum?expand=1#diff-31fd7f1f95597fe1adddc5758f95cc7bfd2ac88aac6476a76cfe71b3a569fd47)

84 new dedicated tests are added to [test/core/test_returncodes.py](https://github.com/jmcnamara/XlsxWriter/compare/main...fra87:returnCodesToEnum?expand=1#diff-a40534b8bb13755f0d1c4860ebdecb0bb4e037c21a282e4ffb65bc503e2c5b2a)

Documentation has been updated

At the end I run make test, make test_flake8 and make docs. All tests succeeded and flake8 did not report anything. Development and tests were done with Python 3.9.2

fra87 commented on xlsxwriter/test/core/test_returncodes.py:
```diff
@@ -0,0 +1,1140 @@
+        # Here we must use an actual file because the worksheet file handle is
+        # not cleaned up correctly when using the StringIO
+
+        filepath = tempfile.mktemp()
```
This is the same as the code in test_close.py (function test_workbook_closes_all_handles); do you think this shall be modified?

jmcnamara replied:
It was probably okay to ignore this warning since it was only in a test case but you fixed it anyway and that is better.

jmcnamara commented:
I'll admit that I wasn't very enthusiastic about this change, and I am worried that it will break existing check on return values, but you clearly but thought and effort into this so you changed my mind.

I do have quite a few comments to make and if you bear with me through those changes then I'll merge the PR.

jmcnamara commented:
You will need to squash the two commits into one.

fra87 commented:
I'm happy to hear that; I originally thought it would have been much easier, but there were lots of dependencies (and I wanted to write a complete set of tests).

In any case let me know if there is anything you think is worth modifying.

I squashed the two commits in one, but then there was a conflict. I fixed it properly (actually it was already fixed in the new commit, because the issue was opened before I started working on the modification), but then the merge commit appeared and I'm not sure I can squash also the merge commit. How can this be solved? Can the two commits be pulled, or do you think it is better to start from scratch on a new branch generated from the new master, copy everything from the old branch in the new and then commit?

fra87 commented:
Closing PR because PR #892 was created (including all the modifications to master up to today)
